### ISSUE TITLE
Struct declarations in C header files cause compilation errors in D modules.

### ISSUE THREAD

**Issue by russel** [jacob-carlborg/dstep#143]
I have stripped this down from a real libdvbv5 problem. We have two C header files, one contains a declaration of a struct, the other the definition of the struct. So in a.h:
```
struct thingy;
```
and in b,h:
```
struct thingy { int i; };
```
Applying dstep we get the D modules as expected. a.d:
```
struct thingy;
```
and b.d
```
struct thingy { int i; }
```
Nothing seemingly wrong so far. But now let us introduce a D module using this type, c.d
```
import a;
import b;

struct blah {
    thingy* x;
}
```
Trying to compile this will lead to:
```
c.d(5): Error: a.thingy at a.d(1) conflicts with b.thingy at b.d(1)
```
I think I (sort of) understand what is happening here, and I feel this has to be considered as a DStep bug.

The struct declaration in a.h – which just states that there is a name of a struct that will be defined elsewhere, to be handled as and when – is becoming in a.d an empty struct definition. Because of the switch from header files to modules a.thingy is not the same as b.thingy in the D modules, whereas in the C header files thingy in a.h is just a declaration of thingy in b.h to keep the compiler happy.  (thingy* only needs to know that thingy is a type name, it doesn't need the structure information, so a simple declaration of thingy is all that is required for C (and C++) codes.)

The problem is that DStep is only taking a local view of all statements in the C code. In the case of a struct (or class) declaration, with no definition, it is not correct to write out what in D becomes a definition.

The problem here clearly relates to Issue #139 . The problem can only be solved by DStep not generating struct and class statements for C struct and class declarations (only for struct and class definitions) and inserting the import statement for the generated module where the struct and class definitions are.

**Comment by ciechowoj** [comment]
This is related to #101 as well.

**Comment by jacob-carlborg** [comment]
I *think* this can be solved with the assumption that all files will be passed to DStep at the same time.

**Comment by Robert-M-Muench** [comment]
The current version still has this problem. I explicitly feed all files to DStep on the command line and had DStep generate separate files.

However, empty struct forward references are still generated. Or is it necessary to just generate one output file?

**Comment by jacob-carlborg** [comment]
> The current version still has this problem.

The bug is still open 😉.

**Comment by Robert-M-Muench** [comment]
Yes,  just wanted to confirm it :-)

Isn't the fix pretty simple? Just don't emit any lines of the form: ```struct mystruct;````

In my case, I commented all those lines and things work.

**Comment by jacob-carlborg** [comment]
> Isn't the fix pretty simple? Just don't emit any lines of the form: ```struct mystruct;````

No. There are cases when no corresponding definition of the struct exists.

**Comment by Robert-M-Muench** [comment]
Well, then it could be cross-checked and only omitted if a struct is not found.
